lib/modem_attest_token/modem_attest_token.c

```diff
@@ -0,0 +1,466 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/at_cmd.h>
+#include <sys/base64.h>
+#include <tinycbor/cbor.h>
+#include <tinycbor/cbor_buf_reader.h>
+#include <modem/modem_attest_token.h>
+
+#define GET_BASE64_LEN(n) (((4 * n / 3) + 3) & ~3)
+#define BASE64_PAD_CHAR '='
+#define AT_ATTEST_CMD "AT%ATTESTTOKEN"
+
+/* Attestation message format:
+ * Tag (55799)
+ * Array (len 5)
+ * - INT: msg type (NRF_ID_SRVC_MSG_TYPE_ID_V1)
+ * - BSTR: device UUID
+ * - INT: device type
+ * - BSTR: FW UUID
+ * - BSTR: nonce
+ */
+#define ATTEST_MAP_LEN 5
+
+#define ATTEST_TOKEN_RESP_LEN 200
+BUILD_ASSERT(
+	ATTEST_TOKEN_RESP_LEN <= CONFIG_AT_CMD_RESPONSE_MAX_LEN,
+	"Max AT cmd response length is too small for attestation token response");
+
+static int parse_resp(char *const resp, char **attest, char **cose)
+{
+	if (!resp || !attest || !cose) {
+		return -EINVAL;
+	}
+
+	char *cose_end = NULL;
+	char *attest_end = NULL;
+
+	*cose = NULL;
+
+	*attest = strchr(resp, '"');
+	if (!(*attest)) {
+		return -EBADMSG;
+	}
+	/* Move beyond first quotation mark */
+	*attest = *attest + 1;
+
+	attest_end = strchr(*attest, '.');
+	if (!(*attest_end)) {
+		return -EBADMSG;
+	}
+
+	/* Replace . with a NULL */
+	*attest_end = '\0';
+
+	*cose = attest_end + 1;
+	cose_end = strchr(*cose, '"');
+	if (!cose_end) {
+		return -EBADMSG;
+	}
+
+	/* Replace " with a NULL */
+	*cose_end = '\0';
+
+	return 0;
+}
+
+int modem_attest_token_get(struct nrf_attestation_token *const token)
+{
+	if (!token) {
+		return -EINVAL;
+	}
+
+	int ret = 0;
+	enum at_cmd_state state;
+	char *cmd_resp = NULL;
+	char *attest = NULL;
+	char *cose = NULL;
+	size_t sz;
+
+	token->attest = NULL;
+	token->cose = NULL;
+
+	/* Init require modules */
+	ret = nrf_modem_lib_init(NORMAL_MODE);
+	if (ret) {
+		return ret;
+	}
+
+	ret = at_cmd_init();
+	if (ret) {
+		return ret;
+	}
+
+	/* Allocate response buffer and send cmd */
+	cmd_resp = k_calloc(ATTEST_TOKEN_RESP_LEN, 1);
+	if (!cmd_resp) {
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Execute AT command to get attestation token */
+	ret = at_cmd_write(AT_ATTEST_CMD, cmd_resp, ATTEST_TOKEN_RESP_LEN,
+			   &state);
+	if (ret) {
+		ret = -EBADMSG;
+		goto cleanup;
+	}
+
+	/* Parse response to get attestation token and COSE strings */
+	ret = parse_resp(cmd_resp, &attest, &cose);
+	if (ret) {
+		ret = -ENOMSG;
+		goto cleanup;
+	}
+
+	/* Allocate and copy attest token string */
+	sz = strlen(attest) + 1;
+	token->attest = k_calloc(sz, 1);
+	if (!token->attest) {
+		ret = -ENOMEM;
+		goto cleanup;
+	} else {
+		memcpy(token->attest, attest, sz);
+	}
+
+	/* Allocate and copy COSE string */
+	sz = strlen(cose) + 1;
+	token->cose = k_calloc(sz, 1);
+	if (!token->cose) {
+		ret = -ENOMEM;
+		if (token->attest) {
```

Comment:
Nit: the check on line 128 makes sure that we don't reach this line if `token->attest` is null